### PR TITLE
Add Multi Information Source Augmented GP

### PR DESCRIPTION
**Pull request by andreaponti5**
## Motivation

This pull request introduces the implementation of the **Augmented Gaussian Process** (AGP) and a related acquisition function based on UCB, namely **Augmented UCB** (AUCB), for multi information source problems.

> Candelieri, A., Archetti, F. Sparsifying to optimize over multiple information sources: an augmented Gaussian process based algorithm. Struct Multidisc Optim 64, 239–255 (2021). [https://doi.org/10.1007/s00158-021-02882-7](https://doi.org/10.1007/s00158-021-02882-7)

### AGP and AUCB in a nutshell
The key idea of the AGP is to fit a GP model for each information source and *augment* the observations on the high fidelity source with those from *cheaper* sources which can be considered as *reliable*. The GP model fitted on this *augmented* set of observations is the AGP.

The AUCB is a modification of the standard UCB - computed on the AGP - suitably proposed to also deal with the source-specific query cost.

### Example
This is what the AGP and AUCB look like on the Forrester problem considering 2 sources.
<p align="center">

</p>

### Some implementation details
The Augmented GP implementation is based on the `SingleTaskGP`. 
Each source is implemented as an independent `SingleTaskGP` and all the _reliable_ observations are used to fit the SingleTaskGP representing the AGP, namely `SingleTaskAugmentedGP`. A key difference with the Multi-Fidelity approaches in BoTorch, is that the dimension representing the source is not directly modelled by the AGP. 
In addition, a Fixed-Noise version of the AGP has been implemented based on the `FixedNoiseGP`.

The Augmented UCB implementation is based on the `UpperConfidenceBound`, but it is penalized by the cost of the source and the discrepancy from the AGP.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/pytorch/botorch/blob/main/CONTRIBUTING.md#pull-requests)?

Yes, I have read the Contributing Guidelines on pull requests.

## Test Plan

A bunch of unit tests for both, the AGP and AUCB, has been implemented. Those are inspired by the `SingleTaskGP` and `UpperConfidenceBound` tests, respectively.

In addition, a notebook tutorial shows how to use the AGP model along with the AUCB acquisition function and a comparison with the Multi-Fidelity MES on the Augmented Branin test function.

Finally, we plan to release an arXiv soon, with an extensive comparison of Multi-Fidelity and Multi Information Source approaches in BoTorch. Here are some preliminary results on the Augmented Hartmann test function, considering three sources (with fidelities $[0.50, 0.75, 1.00]$). The AGP has been compared with the discrete Multi-Fidelity version of the Knowled Gradient (KG), Max-value Entropy Search (MES) and GIBBON. The figure on the left shows the best seen with respect to the query cost, while the figure on the right shows the best seen with respect to the wall-clock time.
<p align="center">

</p>

## Related PRs
The docs has been updated in this PR.

### CODE REVIEW

**Comment by sdaulton** [comment]
Thanks for putting up the PR! This works looks quite interesting!

File: test/models/test_gp_regression_multisource.py [hunk 1]
```diff
@@ -0,0 +1,488 @@
...
+def _get_random_data_with_source(batch_shape, n, d, n_source, q=1, **tkwargs):
+    dtype = tkwargs.get("dtype", torch.float32)
+    rep_shape = batch_shape + torch.Size([1, 1])
+    bounds = torch.stack([torch.zeros(d), torch.ones(d)])
+    bounds[-1, -1] = n_source - 1
+    train_x = (
+        get_random_x_for_agp(n=n, bounds=bounds, q=q).repeat(rep_shape).type(dtype)
+    )
+    train_y = torch.sin(train_x[..., :1] * (2 * math.pi)).type(dtype)
+    train_y = train_y + 0.2 * torch.randn(n, 1, **tkwargs).repeat(rep_shape)
```
Comment (esantorella) [inline test/models/test_gp_regression_multisource.py:44]: When run with `self.device` as CUDA, this is failing with the error "Expected all tensors to be on the same device" because `**tkwargs` puts the random noise on the GPU while the other tensors involved are on the CPU. I think we'd want everything to be on the GPU.

File: test/acquisition/test_multi_source.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+            module = AugmentedUpperConfidenceBound(
+                model=mm,
+                beta=1.0,
+                best_f=torch.tensor(0.5, device=self.device, dtype=dtype),
+                cost={0: 1, 1: 0.5},
+            )
+            X = torch.zeros(1, 2, device=self.device, dtype=dtype)
+            ucb = module(X)
+            ucb_expected = torch.tensor([1.8460], device=self.device, dtype=dtype)
+            self.assertAllClose(ucb, ucb_expected, atol=1e-4)
```
Comment (esantorella) [inline test/acquisition/test_multi_source.py:41]: I'm seeing this fail on a GPU, when `self.device` is "cuda:0", because `ucb` is on the CPU. Looking at other unit tests, it seems like the convention is that `ucb` should be on the CPU.

File: test/models/test_gp_regression_multisource.py (hunk 1 above)
Comment (andreaponti5) [reply test/models/test_gp_regression_multisource.py:43]: Thanks for the comment, this should be solved now. All the tensors are on the same device.

File: test/acquisition/test_multi_source.py (hunk 2 above)
Comment (andreaponti5) [reply test/acquisition/test_multi_source.py:41]: Fixed it. The problem was, as you pointed out, in the computation of `ucb`.

**Comment by saitcakmak** [comment]
Hi @andreaponti5. Thanks for the PR! We're trying out a new support model for community contributions: https://github.com/pytorch/botorch/blob/main/CONTRIBUTING.md#community-contributions-beta. Can you move the new additions to the `botorch_community`, `notebooks_community` & `test_community` directories? Thanks!

File: botorch_community/models/gp_regression_multisource.py [hunk 3]
```diff
@@ -0,0 +1,401 @@
...
+            mean_module=mean_module,
+            outcome_transform=outcome_transform,
+            input_transform=input_transform,
+        )
+        mll = ExactMarginalLogLikelihood(gp.likelihood, gp)
+        fit_gpytorch_mll(mll)
+        return gp
+
+
+class FixedNoiseAugmentedGP(FixedNoiseGP):
```
Comment (Balandat) [inline botorch_community/models/gp_regression_multisource.py:231]: FWIW we have deprecated `FixedNoiseGP` and incorporated its functionality into `SingleTaskGP` by allowing to pass `train_Yvar` also to `SingleTaskGP` (optionally). Would be great to do the same setup here so that the implementations are analogous.

File: sphinx/source/models.rst [hunk 4]
```diff
@@ -182,4 +182,4 @@ Inducing Point Allocators
 Other Utilties
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 .. automodule:: botorch.models.utils.assorted
-    :members:
+    :members:
```
Comment (Balandat) [inline sphinx/source/models.rst:185]: what happened here?

File: botorch_community/models/gp_regression_multisource.py [hunk 5]
```diff
@@ -0,0 +1,401 @@
...
+
+
+class SingleTaskAugmentedGP(SingleTaskGP):
+    r"""A single-task multi-source GP model.
+
+    The Augmented Gaussian Process is described in [Ca2021ms]_.
+    The basic idea is to use GP sparsification for selecting a subset of
+    the function evaluations, among those performed so far over all the
+    different sources, as inducing locations to generate the AGP approximating f(x).
+    The GP sparsification proposed is an insertion method: the set of inducing
```
Comment (Balandat) [inline botorch_community/models/gp_regression_multisource.py:83]: The term "inducing locations" is a bit confusing here I feel, since this is typically used for inducing points in approximate GP models. But IIUC here there aren't any approximate GP models, it's just an adaptive data selection scheme to select which inputs to use in the model (in particular, there is no interpolation or the like going on).

**Review by Balandat — COMMENTED** [review]
Thanks - looks like there are some minor lint and auto-sort code formatting issues. I also left a couple of inline comments that would be great to address (though potentially ok to do this in a follow-up PR).

File: botorch_community/models/gp_regression_multisource.py (hunk 3 above)
Comment (andreaponti5) [reply botorch_community/models/gp_regression_multisource.py]: I incorporated `FixedNoiseAugmentedAGP` in `SingleTaskAugmentedGP` (and consequently the unit tests).

File: sphinx/source/models.rst (hunk 4 above)
Comment (andreaponti5) [reply sphinx/source/models.rst:185]: Not sure what happened. I removed `SingleTaskAugmentedGP` from the doc. Community contributions shouldn't go in the documentation, right?

File: botorch_community/acquisition/augmented_multisource.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+            A `(b1 x ... bk)`-dim tensor of Augmented Upper Confidence Bound values at
+            the given design points `X`.
+        """
+        alpha = torch.zeros(X.shape[0], dtype=X.dtype, device=X.device)
+        agp_mean, agp_sigma = self._mean_and_sigma(X[..., :-1])
+        cb = (self.best_f if self.maximize else -self.best_f) + (
+            (agp_mean if self.maximize else -agp_mean) + self.beta.sqrt() * agp_sigma
+        )
+        source_idxs = {
+            int(s.tolist()): torch.where(torch.round(X[..., -1], decimals=0) == s)[0]
```
[the remark this answers is not in the document]
Comment (Balandat) [reply botorch_community/acquisition/augmented_multisource.py]: TIL that calling `.tolist()` on a zero-dim tensor produces a scalar. `item()` definitely makes more sense.

File: sphinx/source/models.rst (hunk 4 above)
Comment (Balandat) [reply sphinx/source/models.rst:185]: @saitcakmak what are your thoughts re community contribs? Would be nice if we generated docs as part of the community modules (maybe not part of the core modules) given that people are putting effort into their docstrings (thanks!).

Comment (saitcakmak) [reply sphinx/source/models.rst:185]: Yeah, I think it'd be good to have generated docs for community modules on the website. We can build it when building the community tutorials for the website.

File: botorch_community/models/gp_regression_multisource.py (hunk 5 above)
Comment (andreaponti5) [reply botorch_community/models/gp_regression_multisource.py]: Yes, it is based on a GP sparsification technique. We changed the description. I hope it's more clear now.

Comment (Balandat) [reply botorch_community/models/gp_regression_multisource.py]: thanks

**Comment by Balandat** [comment]
Merged! Thanks again @andreaponti5 for your contribution.
